File: Graph Theory Peaks/DFS/Q4. Bus Routes/Q4_BusRoutes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        
        if (source == target) return 0; 
        // If source and target are same → no bus needed

        int n = routes.length; 
        
        Map<Integer, List<Integer>> stopToRoutes = new HashMap<>();
        // Map: busStop -> list of routes (buses) that pass through that stop

        // Build the mapping of stop → routes
        for (int i = 0; i < n; i++) {
            for (int stop : routes[i]) {
                stopToRoutes.computeIfAbsent(stop, x -> new ArrayList<>()).add(i);
                // Add route index to the stop
            }
        }

        Queue<Integer> queue = new LinkedList<>();
        // BFS queue storing route numbers (buses)

        boolean[] visitedRoutes = new boolean[n];
        // Track visited buses

        Set<Integer> visitedStops = new HashSet<>();
        // Track visited stops (not heavily used here but good practice)

        // Start BFS from all buses available at the source stop
        for (int route : stopToRoutes.getOrDefault(source, new ArrayList<>())) {
            queue.offer(route);          // add route to BFS
            visitedRoutes[route] = true; // mark route as visited
        }

        int busesTaken = 1; 
        // First level = taking first bus

        while (!queue.isEmpty()) {
            int size = queue.size(); // number of routes at current BFS level

            for (int i = 0; i < size; i++) {
                int routeIdx = queue.poll(); // current bus route

                // Check all stops of this route
                for (int stop : routes[routeIdx]) {

                    if (stop == target) return busesTaken;
                    // If this route reaches target stop → return buses used

                    // Find other routes connected through this stop
                    for (int neighborRoute : stopToRoutes.getOrDefault(stop, new ArrayList<>())) {

                        // If that route is not visited yet
                        if (!visitedRoutes[neighborRoute]) {
                            visitedRoutes[neighborRoute] = true; 
                            queue.offer(neighborRoute); // take that bus
                        }
                    }
                }
            }

            busesTaken++; 
            // Moving to next level = taking one more bus
        }

        return -1; 
        // Target stop cannot be reached
    }
}

File: Graph Theory Peaks/DFS/Q4. Bus Routes/Q4_BusRoutes.py
from collections import defaultdict, deque

class Solution:
    def numBusesToDestination(self, routes, source, target):

        if source == target:
            return 0  # No bus needed if source and target are same

        stop_to_routes = defaultdict(list)
        # Map stop to list of route indices passing through it

        # Build mapping of stop to routes
        for i, route in enumerate(routes):
            for stop in route:
                stop_to_routes[stop].append(i)

        visited_routes = set()
        # Track buses (routes) already used

        queue = deque()
        # BFS queue storing route indices

        # Start BFS from routes that include the source stop
        for route_idx in stop_to_routes[source]:
            queue.append(route_idx)
            visited_routes.add(route_idx)

        buses_taken = 1
        # First level means first bus taken

        while queue:

            for _ in range(len(queue)):
                route_idx = queue.popleft()
                # Current bus route

                for stop in routes[route_idx]:
                    # Check all stops of this route

                    if stop == target:
                        return buses_taken
                        # Target stop reached

                    # Find other routes connected through this stop
                    for neighbor_route in stop_to_routes[stop]:

                        if neighbor_route not in visited_routes:
                            visited_routes.add(neighbor_route)
                            queue.append(neighbor_route)
                            # Add new bus route to queue

            buses_taken += 1
            # Increase bus count for next BFS level

        return -1
        # Target cannot be reached